portal-wire-protocol.md

```diff
@@ -0,0 +1,163 @@
+message_id      := uint8
+encoded_message := bytes
+```
+
+The `encoded_message` component is the SSZ encoded payload for the message as is indicated by its `message_id`. Each message has its own `sedes` which dictates how it should be encoded and decoded.
+
+The SSZ sedes `ByteList` is used to alias `List[uint8, max_length=2048]`.
+
+All messages have a `type` which is either `request` or `response`.
+
+* `request` messages **MUST** be sent using a `TALKREQ`
+* `response` messages **MUST** be sent using a `TALKRESP`
+
+
+### Ping (0x01)
+
+Request message to check if a node is reachable, communicate basic information about our node, and request basic information about the recipient node.
+
+```
+message_id := 0x01
+type       := request
+sedes      := Container(enr_seq: uint64, data_radius: uint256, custom_payload: ByteList)
+```
+
+* `enr_seq`: The node's current sequence number of their ENR record.
+* `data_radius`: The nodes current maximum radius for data stored by this node.
+* `custom_payload`: Custom payload dependant on the network.
+
+
+### Pong (0x02)
+
+Response message to Ping(0x01)
+
+```
+message_id := 0x02
+type       := response
+sedes      := Container(enr_seq: uint64, data_radius: uint256)
+```
+
+* `enr_seq`: The node's current sequence number of their ENR record.
+* `data_radius`: The nodes current maximum radius for data stored by this node.
+* `custom_payload`: Custom payload dependant on the network.
+
+### Find Nodes (0x03)
+
+Request nodes from the peer's routing table at the given logarithmic distances.  The distance of `0` indicates a request for the peer's own ENR record.
+
+```
+message_id := 0x03
+type       := request
+sedes      := Container(distances: List[uint16, max_length=256])
+```
+
+* `distances`: a list of distances for which the node is requesting ENR records for.
+    * Each distance **MUST** be within the inclusive range `[0, 256]`
+    * Each distance in the list **MUST** be unique.
+
+### Nodes (0x04)
+
+Response message to FindNodes(0x03).
+
+```
+message_id := 0x04
+type       := response
+sedes      := Container(total: uint8, enrs: List[ByteList, max_length=32])
+```
+
+* `total`: The total number of `Nodes` response messages being sent.
+* `enrs`: List of bytestrings, each of which is an RLP encoded ENR record.
+    * Individual ENR records **MUST** correspond to one of the requested distances.
+    * It is invalid to return multiple ENR records for the same `node_id`.
+
+> Note: If the number of ENR records cannot be encoded into a single message, then they should be sent back using multiple messages, with the `total` field representing the total number of messages that are being sent.
+
+### Find Content (0x05)
+
+Request either the data payload for a specific piece of content on the network, **or** ENR records of nodes that are closer to the requested content.
+
+```
+message_id := 0x05
+type       := request
+sedes      := Container(content_key: ByteList)
```

Comment:
This might be a good enough reason for using:

```
ContentKey = Container(type: uint8, payload: ByteList)
```

Reply:
Could you elaborate on this change? Are you suggesting to allow passing the `ContentKey` content type at this level?

Reply:
After spending longer than I'm proud of writing down what I was musing over I'm not convinced it really improves things.  Please disregard.